README.md
# apt-transport-tor

Easily install Debian packages via Tor.

This package implements an APT "acquire method" that handles URLs starting
with "tor+http://" or "tor+https://" in your sources.list.

## Installation

### Via apt

This package will soon be available in Debian:

    apt-get install apt-transport-tor

### From source

If you are working from a git checkout, first run:

    autoreconf -i

Then, or if installing from a tarball:

    ./configure --prefix=/usr
    make
    sudo make install

## Usage

Edit your /etc/apt/sources.list like so, adjusting the suite/components
appropriately for your system:

    deb     tor+http://http.debian.net/debian unstable main
    deb-src tor+http://http.debian.net/debian unstable main

Note the use of http.debian.net so that a mirror close to your exit node
will be automatically chosen.

Alternatively, if you have the Tor hidden service address of a Debian
mirror, you can use that:

    deb     tor+http://<long string>.onion/debian unstable main
    deb-src tor+http://<long string>.onion/debian unstable main

## Configuration

Most users should not need to adjust SOCKS settings.

By default, apt-transport-tor uses the following SOCKS proxy setting, which
matches the default Tor SOCKS port:

    socks5h://apt:apt@localhost:9050

If you want to use a different port, you can edit the Acquire::tor::proxy
apt preference:

    Acquire::tor::proxy "socks5h://apt:apt@localhost:9050";

Note the use of a username/password to make use of the default
IsolateSOCKSAuth Tor setting for stream isolation, which requires bug fixes
from Tor 0.2.4.19 to work well.  This means your apt traffic will be sent
over a different circuit from your regular Tor traffic.

Although "sock5h://" is put explicitly in these examples, at the moment its
use is hardcoded (to avoid DNS leaks).

## Caveats

Downloading your Debian packages over Tor prevents an attacker who is
sniffing your network connection from being able to tell which packages
you are fetching, or even that your traffic is Debian-related.

However, this does not necessarily defend you from, amongst other things:

* a global passive adversary (who could potentially correlate the exit
  node's traffic with your local Tor traffic)
* an attacker looking at the size of your downloads, and making an
  educated guess about the contents
* an attacker who has broken into your machine

Download speeds will be slower via Tor.

## Copyright & Licensing

Copyright (C) 2014 Tim Retout <diocles@debian.org>

apt-transport-tor was forked from the APT https transport.  APT has this
copyright notice:

    Apt is copyright 1997, 1998, 1999 Jason Gunthorpe and others.

License:

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.

## Feedback

Comments and suggestions to: Tim Retout <diocles@debian.org>

Bug reports should be sent to the Debian BTS.


debian/changelog
apt-transport-tor (0.2.1-1) unstable; urgency=medium

  * New version 0.2.1
    + Hardcodes User-Agent string, so all apt-transport-tor users
      will look the same, regardless of the software version.
    + You can now write "tor+http://" instead of "tor://".
    + HTTPS is now also supported, with "tor+https://" URLs.

 -- Tim Retout <diocles@debian.org>  Sun, 20 Jul 2014 14:06:59 +0100

apt-transport-tor (0.1.1-2) unstable; urgency=medium

  * debian/control: Update description to give less of a false sense
    of security.  (Closes: #746773)

 -- Tim Retout <diocles@debian.org>  Sun, 15 Jun 2014 17:29:51 +0100

apt-transport-tor (0.1.1-1) unstable; urgency=medium

  * Initial release. (Closes: #745259)

 -- Tim Retout <diocles@debian.org>  Mon, 21 Apr 2014 20:20:00 +0100


tor.h
// -*- mode: cpp; mode: fold -*-
// Description								/*{{{*/// $Id: http.h,v 1.12 2002/04/18 05:09:38 jgg Exp $
// $Id: http.h,v 1.12 2002/04/18 05:09:38 jgg Exp $
/* ######################################################################

   HTTP Acquire Method - This is the HTTP acquire method for APT.

   ##################################################################### */
									/*}}}*/

#ifndef APT_TOR_H
#define APT_TOR_H

#include <apt-pkg/acquire-method.h>

#include <curl/curl.h>
#include <iostream>
#include <stddef.h>
#include <string>

#include "server.h"

using std::cout;
using std::endl;

class Hashes;
class TorMethod;
class FileFd;

class TorServerState : public ServerState
{
   protected:
   virtual bool ReadHeaderLines(std::string &/*Data*/) { return false; }
   virtual bool LoadNextResponse(bool const /*ToFile*/, FileFd * const /*File*/) { return false; }

   public:
   virtual bool WriteResponse(std::string const &/*Data*/) { return false; }

   /** \brief Transfer the data from the socket */
   virtual bool RunData(FileFd * const /*File*/) { return false; }

   virtual bool Open() { return false; }
   virtual bool IsOpen() { return false; }
   virtual bool Close() { return false; }
   virtual bool InitHashes(FileFd &/*File*/) { return false; }
   virtual Hashes * GetHashes() { return NULL; }
   virtual bool Die(FileFd &/*File*/) { return false; }
   virtual bool Flush(FileFd * const /*File*/) { return false; }
   virtual bool Go(bool /*ToFile*/, FileFd * const /*File*/) { return false; }

   TorServerState(URI Srv, TorMethod *Owner);
   virtual ~TorServerState() {Close();};
};

class TorMethod : public pkgAcqMethod
{
   // minimum speed in bytes/se that triggers download timeout handling
   static const int DL_MIN_SPEED = 10;

   virtual bool Fetch(FetchItem *);
   static size_t parse_header(void *buffer, size_t size, size_t nmemb, void *userp);
   static size_t write_data(void *buffer, size_t size, size_t nmemb, void *userp);
   static int progress_callback(void *clientp, double dltotal, double dlnow, 
				double ultotal, double ulnow);
   void SetupProxy();
   CURL *curl;
   FetchResult Res;
   TorServerState *Server;

   public:
   FileFd *File;
      
   TorMethod() : pkgAcqMethod("1.2",Pipeline | SendConfig), File(NULL)
   {
      File = 0;
      curl = curl_easy_init();
   };

   ~TorMethod()
   {
      curl_easy_cleanup(curl);
   };
};

#include <apt-pkg/strutl.h>
URI Proxy;

#endif


apti18n.h
/* include/apti18n.h.  Generated from apti18n.h.in by configure.  */
// -*- mode: cpp; mode: fold -*-
// $Id: apti18n.h.in,v 1.6 2003/01/11 07:18:18 jgg Exp $
/* Internationalization macros for apt. This header should be included last
   in each C file. */

// Set by autoconf
#define USE_NLS 1

#ifdef USE_NLS
// apt will use the gettext implementation of the C library
#include <libintl.h>
#include <locale.h>
# ifdef APT_DOMAIN
#   define _(x) dgettext(APT_DOMAIN,x)
#   define P_(msg,plural,n) dngettext(APT_DOMAIN,msg,plural,n)
# else
#   define _(x) gettext(x)
#   define P_(msg,plural,n) ngettext(msg,plural,n)
# endif
# define N_(x) x
#else
// apt will not use any gettext
# define setlocale(a, b)
# define textdomain(a)
# define bindtextdomain(a, b)
# define _(x) x
# define P_(msg,plural,n) (n == 1 ? msg : plural)
# define N_(x) x
# define dgettext(d, m) m
#endif


tor.cc
//-*- mode: cpp; mode: fold -*-
// Description								/*{{{*/
// $Id: http.cc,v 1.59 2004/05/08 19:42:35 mdz Exp $
/* ######################################################################

   Tor Acquire Method - This is the Tor acquire method for APT.
   
   It uses libcurl

   ##################################################################### */
									/*}}}*/
// Include Files							/*{{{*/
#include "config.h"

#include <apt-pkg/fileutl.h>
#include <apt-pkg/acquire-method.h>
#include <apt-pkg/error.h>
#include <apt-pkg/hashes.h>
#include <apt-pkg/netrc.h>
#include <apt-pkg/configuration.h>
#include <apt-pkg/macros.h>
#include <apt-pkg/strutl.h>

#include <sys/stat.h>
#include <sys/time.h>
#include <unistd.h>
#include <stdio.h>
#include <iostream>
#include <sstream>
#include <ctype.h>
#include <stdlib.h>

#include "tor.h"

#include <apti18n.h>
									/*}}}*/
using namespace std;

size_t
TorMethod::parse_header(void *buffer, size_t size, size_t nmemb, void *userp)
{
   size_t len = size * nmemb;
   TorMethod *me = (TorMethod *)userp;
   std::string line((char*) buffer, len);
   for (--len; len > 0; --len)
      if (isspace(line[len]) == 0)
      {
	 ++len;
	 break;
      }
   line.erase(len);

   if (line.empty() == true)
   {
      if (me->Server->Result != 416 && me->Server->StartPos != 0)
	 ;
      else if (me->Server->Result == 416 && me->Server->Size == me->File->FileSize())
      {
         me->Server->Result = 200;
	 me->Server->StartPos = me->Server->Size;
      }
      else
	 me->Server->StartPos = 0;

      me->File->Truncate(me->Server->StartPos);
      me->File->Seek(me->Server->StartPos);
   }
   else if (me->Server->HeaderLine(line) == false)
      return 0;

   return size*nmemb;
}

size_t 
TorMethod::write_data(void *buffer, size_t size, size_t nmemb, void *userp)
{
   TorMethod *me = (TorMethod *)userp;

   if (me->Res.Size == 0)
      me->URIStart(me->Res);
   if(me->File->Write(buffer, size*nmemb) != true)
      return false;

   return size*nmemb;
}

int
TorMethod::progress_callback(void *clientp, double dltotal, double /*dlnow*/,
			      double /*ultotal*/, double /*ulnow*/)
{
   TorMethod *me = (TorMethod *)clientp;
   if(dltotal > 0 && me->Res.Size == 0) {
      me->Res.Size = (unsigned long long)dltotal;
   }
   return 0;
}

// TorServerState::TorServerState - Constructor			/*{{{*/
TorServerState::TorServerState(URI Srv,TorMethod * /*Owner*/) : ServerState(Srv, NULL)
{
   TimeOut = _config->FindI("Acquire::tor::Timeout",TimeOut);
   Reset();
}
									/*}}}*/

void TorMethod::SetupProxy()  					/*{{{*/
{
   URI ServerName = Queue->Uri;

   // Curl should never read proxy settings from the environment, as
   // we determine which proxy to use.  Do this for consistency among
   // methods and prevent an environment variable overriding a
   // no-proxy ("DIRECT") setting in apt.conf.
   curl_easy_setopt(curl, CURLOPT_PROXY, "");

   // Determine the proxy setting
   string UseProxy = _config->Find("Acquire::tor::Proxy", _config->Find("Acquire::tor::Proxy").c_str());

   if (UseProxy.empty() == true)
   {
      // Default proxy
      // - socks5h (actually ignored below) - use proxy for DNS resolution
      // - apt:apt@ - dummy socks authentication (for IsolateSOCKSAuth in Tor)
      // - localhost:9050 - default Tor SOCKS port
      UseProxy = "socks5h://apt:apt@localhost:9050";
   }

   // Determine what host and port to use based on the proxy settings
   Proxy = UseProxy;
   if (Proxy.Port != 1)
      curl_easy_setopt(curl, CURLOPT_PROXYPORT, Proxy.Port);
   curl_easy_setopt(curl, CURLOPT_PROXY, Proxy.Host.c_str());
   if (Proxy.User.empty() == false || Proxy.Password.empty() == false)
   {
      curl_easy_setopt(curl, CURLOPT_PROXYUSERNAME, Proxy.User.c_str());
      curl_easy_setopt(curl, CURLOPT_PROXYPASSWORD, Proxy.Password.c_str());
   }

   // Set proxy type to SOCKS5, and let proxy do DNS resolution
   curl_easy_setopt(curl, CURLOPT_PROXYTYPE, CURLPROXY_SOCKS5_HOSTNAME);
}									/*}}}*/
// TorMethod::Fetch - Fetch an item					/*{{{*/
// ---------------------------------------------------------------------
/* This adds an item to the pipeline. We keep the pipeline at a fixed
   depth. */
bool TorMethod::Fetch(FetchItem *Itm)
{
   struct stat SBuf;
   struct curl_slist *headers=NULL;  
   char curl_errorstr[CURL_ERROR_SIZE];
   URI Uri = Itm->Uri;
   string remotehost = Uri.Host;

   // Undo any "tor" or "tor+" at the start
   string prefix="tor+";
   if ("tor" == Uri.Access)
   {
        Uri.Access = "http";
   }
   else if (!Uri.Access.compare(0, prefix.size(), prefix))
   {
        Uri.Access = Uri.Access.substr(prefix.size());
   }

   // TODO:
   //       - http::Pipeline-Depth
   //       - error checking/reporting
   //       - more debug options? (CURLOPT_DEBUGFUNCTION?)

   curl_easy_reset(curl);
   SetupProxy();

   maybe_add_auth (Uri, _config->FindFile("Dir::Etc::netrc"));

   // callbacks
   curl_easy_setopt(curl, CURLOPT_URL, static_cast<string>(Uri).c_str());
   curl_easy_setopt(curl, CURLOPT_HEADERFUNCTION, parse_header);
   curl_easy_setopt(curl, CURLOPT_WRITEHEADER, this);
   curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_data);
   curl_easy_setopt(curl, CURLOPT_WRITEDATA, this);
   curl_easy_setopt(curl, CURLOPT_PROGRESSFUNCTION, progress_callback);
   curl_easy_setopt(curl, CURLOPT_PROGRESSDATA, this);
   // options
   curl_easy_setopt(curl, CURLOPT_NOPROGRESS, false);
   curl_easy_setopt(curl, CURLOPT_FILETIME, true);
   // Allow curl to handle just the protocols we want
   curl_easy_setopt(curl, CURLOPT_PROTOCOLS, CURLPROTO_HTTP | CURLPROTO_HTTPS);
   curl_easy_setopt(curl, CURLOPT_REDIR_PROTOCOLS, CURLPROTO_HTTP | CURLPROTO_HTTPS);

   // SSL parameters are set by default to the common (non mirror-specific) value
   // if available (or a default one) and gets overload by mirror-specific ones.

   // File containing the list of trusted CA.
   string cainfo = _config->Find("Acquire::https::CaInfo","");
   string knob = "Acquire::https::"+remotehost+"::CaInfo";
   cainfo = _config->Find(knob.c_str(),cainfo.c_str());
   if(cainfo.empty() == false)
      curl_easy_setopt(curl, CURLOPT_CAINFO,cainfo.c_str());

   // Check server certificate against previous CA list ...
   bool peer_verify = _config->FindB("Acquire::https::Verify-Peer",true);
   knob = "Acquire::https::" + remotehost + "::Verify-Peer";
   peer_verify = _config->FindB(knob.c_str(), peer_verify);
   curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, peer_verify);

   // ... and hostname against cert CN or subjectAltName
   bool verify = _config->FindB("Acquire::https::Verify-Host",true);
   knob = "Acquire::https::"+remotehost+"::Verify-Host";
   verify = _config->FindB(knob.c_str(),verify);
   int const default_verify = (verify == true) ? 2 : 0;
   curl_easy_setopt(curl, CURLOPT_SSL_VERIFYHOST, default_verify);

   // Also enforce issuer of server certificate using its cert
   string issuercert = _config->Find("Acquire::https::IssuerCert","");
   knob = "Acquire::https::"+remotehost+"::IssuerCert";
   issuercert = _config->Find(knob.c_str(),issuercert.c_str());
   if(issuercert.empty() == false)
      curl_easy_setopt(curl, CURLOPT_ISSUERCERT,issuercert.c_str());

   // For client authentication, certificate file ...
   string pem = _config->Find("Acquire::https::SslCert","");
   knob = "Acquire::https::"+remotehost+"::SslCert";
   pem = _config->Find(knob.c_str(),pem.c_str());
   if(pem.empty() == false)
      curl_easy_setopt(curl, CURLOPT_SSLCERT, pem.c_str());

   // ... and associated key.
   string key = _config->Find("Acquire::https::SslKey","");
   knob = "Acquire::https::"+remotehost+"::SslKey";
   key = _config->Find(knob.c_str(),key.c_str());
   if(key.empty() == false)
      curl_easy_setopt(curl, CURLOPT_SSLKEY, key.c_str());

   // Allow forcing SSL version to SSLv3 or TLSv1 (SSLv2 is not
   // supported by GnuTLS).
   long final_version = CURL_SSLVERSION_DEFAULT;
   string sslversion = _config->Find("Acquire::https::SslForceVersion","");
   knob = "Acquire::https::"+remotehost+"::SslForceVersion";
   sslversion = _config->Find(knob.c_str(),sslversion.c_str());
   if(sslversion == "TLSv1")
     final_version = CURL_SSLVERSION_TLSv1;
   else if(sslversion == "SSLv3")
     final_version = CURL_SSLVERSION_SSLv3;
   curl_easy_setopt(curl, CURLOPT_SSLVERSION, final_version);

   // CRL file
   string crlfile = _config->Find("Acquire::https::CrlFile","");
   knob = "Acquire::https::"+remotehost+"::CrlFile";
   crlfile = _config->Find(knob.c_str(),crlfile.c_str());
   if(crlfile.empty() == false)
      curl_easy_setopt(curl, CURLOPT_CRLFILE, crlfile.c_str());

   // cache-control
   if(_config->FindB("Acquire::tor::No-Cache",
	_config->FindB("Acquire::http::No-Cache",false)) == false)
   {
      // cache enabled
      if (_config->FindB("Acquire::tor::No-Store",
		_config->FindB("Acquire::http::No-Store",false)) == true)
	 headers = curl_slist_append(headers,"Cache-Control: no-store");
      stringstream ss;
      ioprintf(ss, "Cache-Control: max-age=%u", _config->FindI("Acquire::tor::Max-Age",
		_config->FindI("Acquire::http::Max-Age",0)));
      headers = curl_slist_append(headers, ss.str().c_str());
   } else {
      // cache disabled by user
      headers = curl_slist_append(headers, "Cache-Control: no-cache");
      headers = curl_slist_append(headers, "Pragma: no-cache");
   }
   curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);

   // speed limit
   int const dlLimit = _config->FindI("Acquire::tor::Dl-Limit",
		_config->FindI("Acquire::http::Dl-Limit",0))*1024;
   if (dlLimit > 0)
      curl_easy_setopt(curl, CURLOPT_MAX_RECV_SPEED_LARGE, dlLimit);

   // set header
   // Hardcoded so that all apt-transport-tor users look the same.
   curl_easy_setopt(curl, CURLOPT_USERAGENT,
			"Debian APT-CURL/1.0 (0.1)");

   // set timeout
   int const timeout = _config->FindI("Acquire::tor::Timeout",
		_config->FindI("Acquire::http::Timeout",120));
   curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT, timeout);
   //set really low lowspeed timeout (see #497983)
   curl_easy_setopt(curl, CURLOPT_LOW_SPEED_LIMIT, DL_MIN_SPEED);
   curl_easy_setopt(curl, CURLOPT_LOW_SPEED_TIME, timeout);

   // set redirect options and default to 10 redirects
   bool const AllowRedirect = _config->FindB("Acquire::tor::AllowRedirect",
	_config->FindB("Acquire::http::AllowRedirect",true));
   curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, AllowRedirect);
   curl_easy_setopt(curl, CURLOPT_MAXREDIRS, 10);

   // debug
   if(_config->FindB("Debug::Acquire::tor", false))
      curl_easy_setopt(curl, CURLOPT_VERBOSE, true);

   // error handling
   curl_errorstr[0] = '\0';
   curl_easy_setopt(curl, CURLOPT_ERRORBUFFER, curl_errorstr);

   // If we ask for uncompressed files servers might respond with content-
   // negotiation which lets us end up with compressed files we do not support,
   // see 657029, 657560 and co, so if we have no extension on the request
   // ask for text only. As a sidenote: If there is nothing to negotate servers
   // seem to be nice and ignore it.
   if (_config->FindB("Acquire::tor::SendAccept", _config->FindB("Acquire::http::SendAccept", true)) == true)
   {
      size_t const filepos = Itm->Uri.find_last_of('/');
      string const file = Itm->Uri.substr(filepos + 1);
      if (flExtension(file) == file)
	 headers = curl_slist_append(headers, "Accept: text/*");
   }

   // if we have the file send an if-range query with a range header
   if (stat(Itm->DestFile.c_str(),&SBuf) >= 0 && SBuf.st_size > 0)
   {
      char Buf[1000];
      sprintf(Buf, "Range: bytes=%li-", (long) SBuf.st_size);
      headers = curl_slist_append(headers, Buf);
      sprintf(Buf, "If-Range: %s", TimeRFC1123(SBuf.st_mtime).c_str());
      headers = curl_slist_append(headers, Buf);
   }
   else if(Itm->LastModified > 0)
   {
      curl_easy_setopt(curl, CURLOPT_TIMECONDITION, CURL_TIMECOND_IFMODSINCE);
      curl_easy_setopt(curl, CURLOPT_TIMEVALUE, Itm->LastModified);
   }

   // go for it - if the file exists, append on it
   File = new FileFd(Itm->DestFile, FileFd::WriteAny);
   Server = new TorServerState(Itm->Uri, this);

   // keep apt updated
   Res.Filename = Itm->DestFile;

   // get it!
   CURLcode success = curl_easy_perform(curl);

   // If the server returns 200 OK but the If-Modified-Since condition is not
   // met, CURLINFO_CONDITION_UNMET will be set to 1
   long curl_condition_unmet = 0;
   curl_easy_getinfo(curl, CURLINFO_CONDITION_UNMET, &curl_condition_unmet);

   File->Close();
   curl_slist_free_all(headers);

   // cleanup
   if (success != 0)
   {
      _error->Error("%s", curl_errorstr);
      unlink(File->Name().c_str());
      return false;
   }

   // server says file not modified
   if (Server->Result == 304 || curl_condition_unmet == 1)
   {
      unlink(File->Name().c_str());
      Res.IMSHit = true;
      Res.LastModified = Itm->LastModified;
      Res.Size = 0;
      URIDone(Res);
      return true;
   }
   Res.IMSHit = false;

   if (Server->Result != 200 && // OK
	 Server->Result != 206 && // Partial
	 Server->Result != 416) // invalid Range
   {
      char err[255];
      snprintf(err, sizeof(err) - 1, "HttpError%i", Server->Result);
      SetFailReason(err);
      _error->Error("%s", err);
      // unlink, no need keep 401/404 page content in partial/
      unlink(File->Name().c_str());
      return false;
   }

   struct stat resultStat;
   if (unlikely(stat(File->Name().c_str(), &resultStat) != 0))
   {
      _error->Errno("stat", "Unable to access file %s", File->Name().c_str());
      return false;
   }
   Res.Size = resultStat.st_size;

   // invalid range-request
   if (Server->Result == 416)
   {
      unlink(File->Name().c_str());
      Res.Size = 0;
      delete File;
      Redirect(Itm->Uri);
      return true;
   }

   // Timestamp
   curl_easy_getinfo(curl, CURLINFO_FILETIME, &Res.LastModified);
   if (Res.LastModified != -1)
   {
      struct timeval times[2];
      times[0].tv_sec = Res.LastModified;
      times[1].tv_sec = Res.LastModified;
      times[0].tv_usec = times[1].tv_usec = 0;
      utimes(File->Name().c_str(), times);
   }
   else
      Res.LastModified = resultStat.st_mtime;

   // take hashes
   Hashes Hash;
   FileFd Fd(Res.Filename, FileFd::ReadOnly);
   Hash.AddFD(Fd);
   Res.TakeHashes(Hash);

   // keep apt updated
   URIDone(Res);

   // cleanup
   Res.Size = 0;
   delete File;

   return true;
}

int main()
{
   setlocale(LC_ALL, "");

   TorMethod Mth;
   curl_global_init(CURL_GLOBAL_SSL) ;

   return Mth.Run();
}

